src/audio_pixel.cpp

```diff
@@ -0,0 +1,270 @@
+    return (((x2 - x) / (x2 - x1)) * y1) + ((x - x1) / (x2 - x1)) * y2;
+}
+
+double closest_val(double val1, double val2, double target) {
+    return abs(target - val1) <= abs(target - val2) ? val1 : val2;
+}
+
+// ************* audio_pixel_block_t ****************
+
+bool audio_pixel_block_t::flush() const {
+    json j;
+    to_json(j);
+    std::string resource_path = GetResourcePath();
+    std::ofstream ofs;
+    ofs.open(resource_path + "/kiwi-block.json");
+
+    if (!ofs.is_open()){
+        return false;
+    }
+    
+    ofs << std::setw(4) << j << std::endl;
+    return true;
+}
+
+void audio_pixel_block_t::to_json(json& j) const {
+    j[std::to_string(m_pix_per_s)] = m_channel_pixels;
+}
+
+void audio_pixel_block_t::update(std::vector<double>& sample_buffer, int num_channels, int sample_rate) {
+    // constructs an audio pixel block using interleaved sample buffer and num channels
+    // clear the current set of audio pixels 
+    m_channel_pixels.clear();
+    int pixels_per_channel = 0;
+
+    // initalize the m_channel_pixels to have empty audio pixel vectors
+    for (int i = 0; i < num_channels; i++) {
+        std::vector<audio_pixel_t> empty_pixel_vec;
+        m_channel_pixels.push_back(empty_pixel_vec);
+    }
+
+    //calcualte the samples per audio pixel and initalize an audio pixel
+    int samples_per_pixel = sample_rate / m_pix_per_s;
+    m_block_size = ceil((sample_buffer.size()/num_channels)/m_pix_per_s);
+
+    for (int channel= 0; channel < num_channels; channel++) {
+        int collected_samples = 0;
+        audio_pixel_t curr_pixel = audio_pixel_t();
+        pixels_per_channel = 0;
+
+        for (int i = 0; i < sample_buffer.size(); i += num_channels) { 
+            double curr_sample = sample_buffer[i];
+
+            // increment collected samples after creating new audio pixel 
+            collected_samples ++;
+    
+            // update all fields of the current pixeld based on the current sample
+            curr_pixel.m_max = std::max(curr_pixel.m_max, curr_sample);
+            curr_pixel.m_min = std::min(curr_pixel.m_min, curr_sample);
+            curr_pixel.m_rms += (curr_sample * curr_sample);
+
+            if (((collected_samples % samples_per_pixel) == 0) || i == sample_buffer.size() - 1) {
+                // complete the RMS calculation, this method accounts of edge cases (total sample % samples per pixel != 0)
+                curr_pixel.m_rms = sqrt(curr_pixel.m_rms / collected_samples);
+
+                // add the curr pixel to its channel in m_channel_pixels, clear curr_pixel
+                m_channel_pixels[channel].push_back(curr_pixel);
+
+                // set max to equal reaper min, and min to equal reaper max
+                curr_pixel = audio_pixel_t();
+                collected_samples = 0;
+                pixels_per_channel++;
+            }
+        }
+    }
+}
+
+std::vector<std::vector<audio_pixel_t>> audio_pixel_block_t::get_pixels(double t0, double t1) {
```

Comment:
depending on how fast this is, this could be low priority, but we're (maybe unnecessarily) copying of the pixels here, while it could be faster to just pass a reference to them. 

something like this: 
```cpp 

template<typename T>
const std::vector<T> get_view(std::vector<T> const &parent, int start, int end)
{
    // make sure start and  end indices are valid
    auto startptr = v.cbegin() + start;
    auto endptr = v.cbegin() + end;
 
    std::vector<T> vec(startptr, endptr);
    return vec;
}
```

Reply:
would return a const vector which is simply a view of the underlying vector (the block)

Reply:
awesome, I couldn't figure this out but I knew there had to be  a better way thank u